KAFKA-9265: Fix kafka.log.Log instance leak on log deletion

soondenana opened the pull request:
KAFKA-8448 fixes problem with similar leak. The Log objects are being
held in ScheduledExecutor PeriodicProducerExpirationCheck callback. The
fix in KAFKA-8448 was to change the policy of ScheduledExecutor to
remove the scheduled task when it gets canceled (by calling
setRemoveOnCancelPolicy(true)).

This works when a log is closed using close() method. But when a log is
deleted either when the topic gets deleted or when the rebalancing
operation moves the replica away from broker, the delete() operation is
invoked. Log.delete() doesn't close the pending scheduled task and that
leaks Log instance.

Fix is to close the scheduled task in the Log.delete() method too.

Tested with and without this fix and Log instances are no longer leaked
and scheduled tasks are gone once log is deleted.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
KAFKA-9265: Fix kafka.log.Log instance leak on log deletion

KAFKA-8448 fixes problem with similar leak. The Log objects are being
held in ScheduledExecutor PeriodicProducerExpirationCheck callback. The
fix in KAFKA-8448 was to change the policy of ScheduledExecutor to
remove the scheduled task when it gets canceled (by calling
setRemoveOnCancelPolicy(true)).

This works when a log is closed using close() method. But when a log is
deleted either when the topic gets deleted or when the rebalancing
operation moves the replica away from broker, the delete() operation is
invoked. Log.delete() doesn't close the pending scheduled task and that
leaks Log instance.

Fix is to close the scheduled task in the Log.delete() method too.

Tested with and without this fix and Log instances are no longer leaked
and scheduled tasks are gone once log is deleted.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -2005,6 +2005,7 @@ class Log(@volatile var dir: File,
       lock synchronized {
         checkIfMemoryMappedBufferClosed()
         removeLogMetrics()
+        producerExpireCheck.cancel(true)
```
It's a little surprising to me that we don't _always_ call `close()` on the `Log` object. Perhaps we can add a few comments which emphasize that the log is considered closed after deletion and no further cleanup is needed. Even better would be if we could consolidate these two paths somehow so that we do not have this kind of bug in the future.

ijuma replied:
Maybe we can do the simple thing in this PR so that we can backport.

hachikuji replied:
I'm happy with that if improvements are not straightforward. Just whenever we punt these tech debts, they tend to never get done.

soondenana replied:
Done.

soondenana replied:
As discussed, limited this PR to only leak fix.

soondenana committed:
make delete call close

This change updates delete to call close so that the functionality gets
reused. Otherwise both close and delete were doing different things and
missing on what one or other does. This change consolidates that logic.

hachikuji commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -317,8 +317,14 @@ abstract class AbstractIndex(@volatile var file: File, val baseOffset: Long, val
-    try ByteBufferUnmapper.unmap(file.getAbsolutePath, mmap)
-    finally mmap = null // Accessing unmapped mmap crashes JVM by SEGV so we null it out to be safe
+    try {
+      // Only mmap if mapped. If its already unmapped do nothing
```
nit: `Only unmap`

soondenana replied:
reverted.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -935,6 +935,23 @@ class Log(@volatile var dir: File,
+  /**
+   * Closes this log after taking the snapshot of state of this log (epochs, offsets etc).
+   */
+  def closeWithProducerStateSnapshot(): Unit = {
```
We have a method called `takeSnapshot` already. Perhaps it would be ok to modify that and call it directly instead of trying to combine the two operations:
```scala
log.flush()
log.takeProducerSnapshot()
log.close()	          
```

soondenana replied:
reverted.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -2004,7 +2018,7 @@ class Log(@volatile var dir: File,
       lock synchronized {
         checkIfMemoryMappedBufferClosed()
-        removeLogMetrics()
+        close()
```
I think we only have one caller of `delete` which is already in a `try` block. Maybe we can let it call `close()` explicitly in the `finally`?

soondenana replied:
reverted.

hachikuji reviewed (commented):
Thanks for the updates. Left a few small suggestions.

soondenana committed:
Revert "make delete call close"

This reverts commit c50b61c0fe31f0cee68002baf9511873bf636ca7.

soondenana committed:
Log.close() now calls removeLogMetrics

This makes both delete and close have same cleanup steps and remove both
users (metrics and PeriodicProducerExpirationCheck callback) of the Log
once log is either closed or deleted.

soondenana committed:
Revert "Log.close() now calls removeLogMetrics"

This reverts commit 30abf65b2fe533ea51ae9456005d6d73a4a4583c.

hachikuji reviewed (approved):
LGTM. Thanks for the fix!

ijuma reviewed (approved):
LGTM

ijuma commented:
retest this please

ijuma commented:
1 job passed and 2 failed with known flakes:

>  kafka.api.SaslOAuthBearerSslEndToEndAuthorizationTest > testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl
> kafka.api.CustomQuotaCallbackTest > testCustomQuotaCallback
> kafka.network.DynamicConnectionQuotaTest > testDynamicConnectionQuota

ijuma commented:
Merged to trunk and cherry-picked to 2.4 and 2.3 branches.
